File: packages/setpieces-r/setpieces/R/events.R
# Extract structured set-piece events (corners, free kicks, throw-ins,
# penalties) out of a raw Opta match object.

#' Extract set-piece events from a match
#'
#' Each returned record has:
#' \itemize{
#'   \item `kind`: `"corner"`, `"free_kick"`, `"throw_in"`, or `"penalty"`
#'   \item `subtype`: `"delivery"` for corner/free_kick/throw_in deliveries;
#'     `"direct_shot_<goal|saved|post|miss>"` for direct free-kick shots;
#'     `"<goal|saved|post|miss>"` for penalties
#'   \item `team_id`, `player_name`, `period`, `minute`, `x`, `y`,
#'     `end_x`, `end_y`, `outcome`
#'   \item `event_index`: this event's position in [sorted_events()] for
#'     the same match -- needed to scan forward for second-ball contests
#' }
#'
#' @param match A parsed match object (from [load_match()]).
#' @param kinds Optional character vector restricting to a subset of
#'   `c("corner", "free_kick", "throw_in", "penalty")`.
#' @return A list of set-piece event records (each a named list).
#' @export
extract_set_pieces <- function(match, kinds = NULL) {
  events <- sorted_events(match)
  out <- list()

  for (i in seq_along(events)) {
    e <- events[[i]]
    type_id <- e$typeId
    if (is.null(type_id)) next
    qids <- qualifier_ids(e)

    kind <- NA_character_
    subtype <- NA_character_

    if (type_id == TYPE_PASS) {
      if (QUALIFIER_CORNER %in% qids) {
        kind <- "corner"; subtype <- "delivery"
      } else if (QUALIFIER_FREE_KICK %in% qids) {
        kind <- "free_kick"; subtype <- "delivery"
      } else if (QUALIFIER_THROW_IN %in% qids) {
        kind <- "throw_in"; subtype <- "delivery"
      }
    } else if (type_id %in% SHOT_TYPES) {
      shot_outcome <- unname(SHOT_OUTCOME_NAMES[as.character(type_id)])
      if (QUALIFIER_PENALTY %in% qids) {
        kind <- "penalty"; subtype <- shot_outcome
      } else if (QUALIFIER_FREE_KICK %in% qids) {
        kind <- "free_kick"; subtype <- paste0("direct_shot_", shot_outcome)
      }
    }

    if (is.na(kind)) next
    if (!is.null(kinds) && !(kind %in% kinds)) next

    if (identical(subtype, "delivery")) {
      end <- pass_end_xy(e)
    } else {
      end <- list(x = NA_real_, y = NA_real_)
    }

    record <- list(
      kind = kind, subtype = subtype,
      team_id = e$contestantId,
      player_name = if (is.null(e$playerName)) NA_character_ else e$playerName,
      period = as.integer(e$periodId), minute = minute_value(e),
      x = as.numeric(e$x), y = as.numeric(e$y),
      end_x = end$x, end_y = end$y,
      outcome = if (is.null(e$outcome)) NA_integer_ else as.integer(e$outcome),
      event_index = i
    )
    class(record) <- "setpiece_event"
    out[[length(out) + 1]] <- record
  }

  class(out) <- c("setpiece_events", "list")
  out
}

#' Convert a list of set-piece events to a data.frame
#'
#' @param events A list of set-piece event records (from
#'   [extract_set_pieces()]).
#' @return A data.frame with one row per event.
#' @export
set_pieces_to_df <- function(events) {
  if (length(events) == 0) {
    return(data.frame(
      kind = character(0), subtype = character(0), team_id = character(0),
      player_name = character(0), period = integer(0), minute = numeric(0),
      x = numeric(0), y = numeric(0), end_x = numeric(0), end_y = numeric(0),
      outcome = integer(0), event_index = integer(0),
      stringsAsFactors = FALSE
    ))
  }
  rows <- lapply(events, function(e) {
    data.frame(
      kind = e$kind, subtype = e$subtype, team_id = e$team_id,
      player_name = e$player_name, period = e$period, minute = e$minute,
      x = e$x, y = e$y, end_x = e$end_x, end_y = e$end_y,
      outcome = e$outcome, event_index = e$event_index,
      stringsAsFactors = FALSE
    )
  })
  do.call(rbind, rows)
}

File: packages/setpieces-py/src/setpieces/events.py
"""Extract structured set-piece events (corners, free kicks, throw-ins,
penalties) out of a raw Opta match dict."""

from dataclasses import dataclass, asdict
from typing import List, Optional

from . import codes
from .geometry import minute_value, pass_end_xy, qualifier_ids
from .io import sorted_events


@dataclass(frozen=True)
class SetPieceEvent:
    """One set-piece delivery or shot.

    kind: "corner" | "free_kick" | "throw_in" | "penalty"
    subtype: "delivery" for corner/free_kick/throw_in deliveries;
        "direct_shot_<goal|saved|post|miss>" for direct free-kick shots;
        "<goal|saved|post|miss>" for penalties.
    event_index: this event's position in ``sorted_events(match)`` --
        needed to scan forward for second-ball contests.
    """

    kind: str
    subtype: str
    team_id: str
    player_name: Optional[str]
    period: int
    minute: float
    x: float
    y: float
    end_x: Optional[float]
    end_y: Optional[float]
    outcome: Optional[int]
    event_index: int

    def as_dict(self) -> dict:
        return asdict(self)


def extract_set_pieces(match: dict, kinds: Optional[List[str]] = None) -> List[SetPieceEvent]:
    """All set-piece events in a match, in chronological order.

    ``kinds`` optionally restricts to a subset of
    {"corner", "free_kick", "throw_in", "penalty"}.
    """
    events = sorted_events(match)
    out: List[SetPieceEvent] = []

    for i, e in enumerate(events):
        type_id = e.get("typeId")
        qids = qualifier_ids(e)
        kind = subtype = None

        if type_id == codes.TYPE_PASS:
            if codes.QUALIFIER_CORNER in qids:
                kind, subtype = "corner", "delivery"
            elif codes.QUALIFIER_FREE_KICK in qids:
                kind, subtype = "free_kick", "delivery"
            elif codes.QUALIFIER_THROW_IN in qids:
                kind, subtype = "throw_in", "delivery"
        elif type_id in codes.SHOT_TYPES:
            shot_outcome = codes.SHOT_OUTCOME_NAMES[type_id]
            if codes.QUALIFIER_PENALTY in qids:
                kind, subtype = "penalty", shot_outcome
            elif codes.QUALIFIER_FREE_KICK in qids:
                kind, subtype = "free_kick", f"direct_shot_{shot_outcome}"

        if kind is None:
            continue
        if kinds is not None and kind not in kinds:
            continue

        end_x, end_y = pass_end_xy(e) if subtype == "delivery" else (None, None)
        out.append(SetPieceEvent(
            kind=kind, subtype=subtype,
            team_id=e.get("contestantId"), player_name=e.get("playerName"),
            period=e.get("periodId"), minute=minute_value(e),
            x=float(e["x"]), y=float(e["y"]),
            end_x=end_x, end_y=end_y,
            outcome=e.get("outcome"),
            event_index=i,
        ))

    return out
